File: .agents/skills/vector-memory/memory/sync_instincts.py
import os
import re
from pathlib import Path
from typing import Dict, Any, List
from .config import PROJECT_TABLE
from .save import save

CONFIDENCE_THRESHOLD = 0.70

def get_homunculus_dirs() -> List[Path]:
    """Find all candidate ECC homunculus directories across Windows / Linux / macOS."""
    candidates = []
    
    # 1. Explicit env overrides
    env_dir = os.environ.get("CLV2_HOMUNCULUS_DIR") or os.environ.get("HOMUNCULUS_DIR")
    if env_dir:
        candidates.append(Path(env_dir))

    # 2. Windows LocalAppData
    appdata = os.environ.get("LOCALAPPDATA")
    if appdata:
        candidates.append(Path(appdata) / "ecc-homunculus")

    # 3. XDG standard path
    xdg = os.environ.get("XDG_DATA_HOME")
    if xdg:
        candidates.append(Path(xdg) / "ecc-homunculus")

    # 4. Standard ~/.local/share path
    candidates.append(Path.home() / ".local" / "share" / "ecc-homunculus")

    # Return deduplicated paths
    seen = set()
    result = []
    for c in candidates:
        norm = str(c.resolve()) if c.exists() else str(c)
        if norm not in seen:
            seen.add(norm)
            result.append(c)
    return result

def get_homunculus_dir() -> Path:
    """Find the primary ECC homunculus directory across Windows / Linux / macOS."""
    dirs = get_homunculus_dirs()
    for d in dirs:
        if d.exists():
            return d
    # If none exist, return the primary candidate and ensure it is created
    primary = dirs[0] if dirs else Path.home() / ".local" / "share" / "ecc-homunculus"
    primary.mkdir(parents=True, exist_ok=True)
    return primary

def parse_simple_yaml_frontmatter(content: str) -> Dict[str, Any]:
    """
    Parse basic YAML frontmatter without external pyyaml dependency.
    """
    data: Dict[str, Any] = {}
    lines = content.splitlines()

    in_frontmatter = False
    fm_lines = []

    for line in lines:
        if line.strip() == "---":
            if not in_frontmatter:
                in_frontmatter = True
                continue
            else:
                in_frontmatter = False
                break
        if in_frontmatter:
            fm_lines.append(line)

    for line in fm_lines:
        if ":" in line:
            key, val = line.split(":", 1)
            k = key.strip()
            v = val.strip().strip("\"'")
            try:
                if "." in v:
                    data[k] = float(v)
                else:
                    data[k] = int(v)
            except ValueError:
                data[k] = v

    return data

def discover_instinct_files(project_id: str = "") -> List[Path]:
    """
    Discover all instinct YAML/YML files across both global and project-scoped directories.
    """
    homunculus_dirs = get_homunculus_dirs()
    found_files: List[Path] = []
    seen_stems = set()

    for hdir in homunculus_dirs:
        if not hdir.exists():
            continue

        # 1. Global instincts: <homunculus>/instincts/personal
        global_personal = hdir / "instincts" / "personal"
        if global_personal.exists():
            for yf in list(global_personal.glob("*.yaml")) + list(global_personal.glob("*.yml")):
                if yf.stem not in seen_stems:
                    seen_stems.add(yf.stem)
                    found_files.append(yf)

        # 2. Project-scoped instincts: <homunculus>/projects/<pid>/instincts/personal
        projects_dir = hdir / "projects"
        if projects_dir.exists():
            if project_id:
                target_dirs = [projects_dir / project_id / "instincts" / "personal"]
            else:
                target_dirs = list(projects_dir.glob("*/instincts/personal"))

            for p_dir in target_dirs:
                if p_dir.exists():
                    for yf in list(p_dir.glob("*.yaml")) + list(p_dir.glob("*.yml")):
                        if yf.stem not in seen_stems:
                            seen_stems.add(yf.stem)
                            found_files.append(yf)

    return found_files

def sync_instincts(threshold: float = CONFIDENCE_THRESHOLD, table: str = PROJECT_TABLE, project_id: str = "") -> Dict[str, Any]:
    """
    Evaluate instincts across global and project-scoped homunculus stores
    and promote those meeting the confidence threshold to LanceDB.
    """
    primary_dir = get_homunculus_dir()
    yaml_files = discover_instinct_files(project_id=project_id)

    results = {
        "homunculus_dir": str(primary_dir),
        "threshold": threshold,
        "scanned": len(yaml_files),
        "promoted": 0,
        "held": 0,
        "details": [],
    }

    if not yaml_files:
        results["message"] = f"No instinct files found across candidate homunculus directories."
        return results

    for yf in yaml_files:
        try:
            content = yf.read_text(encoding="utf-8")
            meta = parse_simple_yaml_frontmatter(content)
            confidence = float(meta.get("confidence", 0.5))
            instinct_id = meta.get("id") or yf.stem
            title = meta.get("name") or meta.get("trigger") or instinct_id
            scope = meta.get("scope", "project")
            pid = meta.get("project_id", "")

            if confidence >= threshold:
                save(
                    title=f"Instinct: {title}",
                    content=content,
                    table=table,
                    category="procedural",
                    metadata={
                        "memory_type": "procedural",
                        "confidence": confidence,
                        "source": "clv2-instinct",
                        "instinct_id": instinct_id,
                        "scope": scope,
                        "project_id": pid,
                        "trust_state": "confirmed",
                        "file": yf.name,
                    },
                    id=f"instinct-{instinct_id}",
                )
                results["promoted"] += 1
                results["details"].append({
                    "id": instinct_id,
                    "confidence": confidence,
                    "scope": scope,
                    "status": "promoted",
                })
            else:
                results["held"] += 1
                results["details"].append({
                    "id": instinct_id,
                    "confidence": confidence,
                    "scope": scope,
                    "status": "held",
                })
        except Exception as err:
            results["details"].append({
                "file": yf.name,
                "status": "error",
                "error": str(err),
            })

    return results

File: .agents/skills/vector-memory/scripts/sync-instincts.js
#!/usr/bin/env node

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { save, getTable, resolveTableName, TABLE_NAMES } from '../src/index.js';

const __filename = fileURLToPath(import.meta.url);

/**
 * Resolve the root data directory for continuous-learning-v2 (ecc-homunculus).
 */
export function resolveHomunculusDir(overrideDir = null) {
  if (overrideDir && fs.existsSync(overrideDir)) {
    return overrideDir;
  }
  if (process.env.CLV2_HOMUNCULUS_DIR && fs.existsSync(process.env.CLV2_HOMUNCULUS_DIR)) {
    return process.env.CLV2_HOMUNCULUS_DIR;
  }
  if (process.env.XDG_DATA_HOME) {
    const p = path.join(process.env.XDG_DATA_HOME, 'ecc-homunculus');
    if (fs.existsSync(p)) return p;
  }
  const home = process.env.USERPROFILE || process.env.HOME || '';
  return path.join(home, '.local', 'share', 'ecc-homunculus');
}

/**
 * Parse an instinct YAML file with markdown body without external dependencies.
 * @param {string} rawContent
 * @returns {Object|null}
 */
export function parseInstinctFile(rawContent) {
  if (!rawContent.startsWith('---')) return null;

  const parts = rawContent.split(/^---$/m);
  if (parts.length < 3) return null;

  const frontmatterStr = parts[1];
  const bodyStr = parts.slice(2).join('---').trim();

  const metadata = {};
  for (const line of frontmatterStr.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const colonIdx = trimmed.indexOf(':');
    if (colonIdx === -1) continue;

    const key = trimmed.slice(0, colonIdx).trim();
    let val = trimmed.slice(colonIdx + 1).trim();

    // Strip surrounding quotes
    if ((val.startsWith('"') && val.endsWith('"')) || (val.startsWith("'") && val.endsWith("'"))) {
      val = val.slice(1, -1);
    }

    if (key === 'confidence') {
      metadata[key] = parseFloat(val);
    } else {
      metadata[key] = val;
    }
  }

  // Extract title from body (# Title)
  let title = metadata.id || 'Untitled Instinct';
  const titleMatch = bodyStr.match(/^#\s+(.+)$/m);
  if (titleMatch) {
    title = titleMatch[1].trim();
  }

  // Extract action section (## Action)
  let action = '';
  const actionMatch = bodyStr.match(/##\s+Action\s*\n([\s\S]*?)(?=\n##|$)/i);
  if (actionMatch) {
    action = actionMatch[1].trim();
  }

  // Extract evidence section (## Evidence)
  let evidence = '';
  const evidenceMatch = bodyStr.match(/##\s+Evidence\s*\n([\s\S]*?)(?=\n##|$)/i);
  if (evidenceMatch) {
    evidence = evidenceMatch[1].trim();
  }

  const content = [
    `Trigger: ${metadata.trigger || 'General'}`,
    action ? `Action:\n${action}` : '',
    evidence ? `Evidence:\n${evidence}` : '',
  ]
    .filter(Boolean)
    .join('\n\n');

  return {
    id: metadata.id || `instinct-${Date.now()}`,
    title,
    content,
    confidence: typeof metadata.confidence === 'number' ? metadata.confidence : 0.5,
    domain: metadata.domain || 'workflow',
    trigger: metadata.trigger || '',
    scope: metadata.scope || 'project',
    projectId: metadata.project_id || null,
    projectName: metadata.project_name || null,
    rawBody: bodyStr,
  };
}

/**
 * Scan directory recursively for .yaml instinct files.
 * @param {string} dir
 * @returns {Array<string>}
 */
function findInstinctFiles(dir) {
  const results = [];
  if (!fs.existsSync(dir)) return results;

  const entries = fs.readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      results.push(...findInstinctFiles(fullPath));
    } else if (entry.isFile() && (entry.name.endsWith('.yaml') || entry.name.endsWith('.yml'))) {
      results.push(fullPath);
    }
  }
  return results;
}

/**
 * Sync instincts through the Approach 2 Promotion / Trust Gate into LanceDB.
 * @param {Object} options
 * @param {number} [options.minConfidence=0.7] - Confidence threshold for promotion
 * @param {boolean} [options.dryRun=false] - Preview without saving to LanceDB
 * @param {string} [options.table] - Target table override
 * @param {string} [options.homunculusDir] - Custom homunculus directory
 * @returns {Promise<Object>} Summary statistics
 */
export async function syncInstincts(options = {}) {
  const {
    minConfidence = 0.7,
    dryRun = false,
    table: tableOverride = null,
    homunculusDir: customDir = null,
  } = options;

  const baseDir = resolveHomunculusDir(customDir);
  console.log(`\n=== Approach 2 Promotion & Trust Gate (sync-instincts) ===`);
  console.log(`Homunculus Source: ${baseDir}`);
  console.log(`Promotion Threshold: confidence >= ${minConfidence}`);
  console.log(`Dry Run Mode:        ${dryRun ? 'YES (No writes)' : 'NO (Writing to LanceDB)'}\n`);

  if (!fs.existsSync(baseDir)) {
    console.log(`Source directory not found (${baseDir}). No instincts to sync yet.`);
    return { scanned: 0, promoted: 0, skipped: 0, sourceDir: baseDir };
  }

  // Scan both global instincts and project-scoped instincts
  const instinctFiles = findInstinctFiles(baseDir);
  console.log(`Found ${instinctFiles.length} instinct file(s) across repositories.\n`);

  let scanned = 0;
  let promoted = 0;
  let skipped = 0;

  for (const filePath of instinctFiles) {
    scanned++;
    const rawContent = fs.readFileSync(filePath, 'utf8');
    const instinct = parseInstinctFile(rawContent);

    if (!instinct) {
      console.log(`[SKIP] Could not parse: ${path.basename(filePath)}`);
      skipped++;
      continue;
    }

    const passesGate = instinct.confidence >= minConfidence;
    const targetTable = tableOverride
      ? resolveTableName(tableOverride)
      : instinct.scope === 'global'
      ? TABLE_NAMES.DEFAULT
      : TABLE_NAMES.PROJECT;

    if (!passesGate) {
      console.log(
        `[GATE: REJECTED] "${instinct.title}" (${instinct.id}) - Confidence ${instinct.confidence} < ${minConfidence}`
      );
      skipped++;
      continue;
    }

    console.log(
      `[GATE: PROMOTED] "${instinct.title}" (${instinct.id})` +
        `\n  Confidence: ${instinct.confidence} >= ${minConfidence} | Scope: ${instinct.scope} | Table: ${targetTable}`
    );

    if (!dryRun) {
      const recordId = `instinct-${instinct.id}`;
      await save({
        table: targetTable,
        id: recordId,
        title: instinct.title,
        content: instinct.content,
        category: 'procedural', // 3-Category Taxonomy: Procedural
        metadata: {
          memory_type: 'procedural',
          trust_state: 'confirmed',
          source: 'continuous-learning-v2',
          instinct_id: instinct.id,
          confidence: instinct.confidence,
          domain: instinct.domain,
          trigger: instinct.trigger,
          scope: instinct.scope,
          project_id: instinct.projectId,
          promoted_at: new Date().toISOString(),
        },
      });
      console.log(`  -> Successfully indexed into ${targetTable} [ID: ${recordId}]`);
    }

    promoted++;
  }

  console.log('\n---------------------------------------------------------');
  console.log(`Sync Complete: ${scanned} scanned, ${promoted} promoted, ${skipped} held at gate.`);
  console.log('---------------------------------------------------------\n');

  return { scanned, promoted, skipped, sourceDir: baseDir };
}

// CLI Execution Handler
if (process.argv[1] === __filename) {
  const args = process.argv.slice(2);
  const dryRun = args.includes('--dry-run') || args.includes('-n');

  const confIdx = args.findIndex((a) => a === '--min-confidence' || a === '-c');
  const minConfidence = confIdx !== -1 ? parseFloat(args[confIdx + 1]) : 0.7;

  const tblIdx = args.findIndex((a) => a === '--table' || a === '-t');
  const table = tblIdx !== -1 ? args[tblIdx + 1] : null;

  const dirIdx = args.findIndex((a) => a === '--dir' || a === '-d');
  const homunculusDir = dirIdx !== -1 ? args[dirIdx + 1] : null;

  syncInstincts({ minConfidence, dryRun, table, homunculusDir }).catch((err) => {
    console.error(`\nSync Failed: ${err.message}\n`, err);
    process.exit(1);
  });
}

export default syncInstincts;
